[iOS] Implement PlaybackSessionInterfaceAVKit in terms of AVMediaSource

aestes opened the pull request:
#### 7c47403ad6ee2fbedfbcbf2d713f0c6c9d45dee1
<pre>
[iOS] Implement PlaybackSessionInterfaceAVKit in terms of AVMediaSource
<a href="https://bugs.webkit.org/show_bug.cgi?id=284799">https://bugs.webkit.org/show_bug.cgi?id=284799</a>
<a href="https://rdar.apple.com/problem/141601785">rdar://problem/141601785</a>

Reviewed by Eric Carlson.

Implemented PlaybackSessionInterfaceAVKit in terms of a content source object (WebAVContentSource)
that conforms to AVMediaSource. AVMediaSource declares a number of read-only properties that are
expected to be key-value observable, and WebAVContentSource conforms by declaring readwrite
properties that are set by various PlaybackSessionInterfaceAVKit functions. AVMediaSource methods
(e.g., -play) are implemented by calling into PlaybackSessionModel.

To avoid duplicating the method from PlaybackSessionInterfaceAVKitLegacy that converts a TimeRanges
to an NSArray of NSValues, a helper function was added to PlatformTimeRanges.

* Source/WebCore/PAL/pal/spi/cocoa/AVKitSPI.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/PlatformTimeRanges.h:
* Source/WebCore/platform/graphics/cocoa/PlatformTimeRangesCocoa.mm: Added.
(WebCore::makeNSArray):
* Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.mm:
(-[WebAVListItem initWithLocalizedTitle:]):
(-[WebAVContentSource initWithModel:]):
(-[WebAVContentSource beginScanningBackward]):
(-[WebAVContentSource beginScanningForward]):
(-[WebAVContentSource endScanningBackward]):
(-[WebAVContentSource endScanningForward]):
(-[WebAVContentSource beginScrubbing]):
(-[WebAVContentSource endScrubbing]):
(-[WebAVContentSource pause]):
(-[WebAVContentSource play]):
(-[WebAVContentSource seekTo:]):
(-[WebAVContentSource setCaptionContentInsets:]):
(-[WebAVContentSource updateCurrentAudioOption:]):
(-[WebAVContentSource updateCurrentCaptionOption:]):
(-[WebAVContentSource updateMuted:]):
(-[WebAVContentSource updateVolume:]):
(WebCore::nowPlayingMetadataObserver):
(WebCore::PlaybackSessionInterfaceAVKit::PlaybackSessionInterfaceAVKit):
(WebCore::PlaybackSessionInterfaceAVKit::durationChanged):
(WebCore::PlaybackSessionInterfaceAVKit::currentTimeChanged):
(WebCore::PlaybackSessionInterfaceAVKit::rateChanged):
(WebCore::PlaybackSessionInterfaceAVKit::seekableRangesChanged):
(WebCore::PlaybackSessionInterfaceAVKit::canPlayFastReverseChanged):
(WebCore::PlaybackSessionInterfaceAVKit::audioMediaSelectionOptionsChanged):
(WebCore::PlaybackSessionInterfaceAVKit::legibleMediaSelectionOptionsChanged):
(WebCore::PlaybackSessionInterfaceAVKit::audioMediaSelectionIndexChanged):
(WebCore::PlaybackSessionInterfaceAVKit::legibleMediaSelectionIndexChanged):
(WebCore::PlaybackSessionInterfaceAVKit::mutedChanged):
(WebCore::PlaybackSessionInterfaceAVKit::volumeChanged):
(WebCore::PlaybackSessionInterfaceAVKit::startObservingNowPlayingMetadata):
(WebCore::PlaybackSessionInterfaceAVKit::stopObservingNowPlayingMetadata):
(WebCore::PlaybackSessionInterfaceAVKit::nowPlayingMetadataChanged):
* Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKitLegacy.mm:
(WebCore::PlaybackSessionInterfaceAVKitLegacy::seekableRangesChanged):

Canonical link: <a href="https://commits.webkit.org/287992@main">https://commits.webkit.org/287992@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7414f53f97801711770af8b4c9734525371a8b31

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81559 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1085 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35507 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86091 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32553 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83669 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1104 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8903 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63644 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21380 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84628 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/784 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74224 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43935 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/684 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28404 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31006 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72116 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28997 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87527 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8792 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6238 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71976 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8973 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70052 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71209 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17733 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15275 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14190 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8752 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14280 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8590 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12112 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10397 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.mm:
```diff
@@ -51,6 +233,100 @@
 
+void PlaybackSessionInterfaceAVKit::durationChanged(double duration)
+{
+    [m_contentSource setMinValue:0];
```
Is 0 always the right minimum value, e.g. for a live stream?

aestes replied:
Great question! I'm not sure; I was copying what we did in `PlaybackSessionInterfaceLMK`. I'll add a FIXME for now.

eric-carlson commented on Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.mm:
```diff
@@ -28,8 +28,178 @@
+}
+
+- (void)setCaptionContentInsets:(UIEdgeInsets)insets
+{
```
Is it worth logging an error here so we can see if it is called?

aestes replied:
I'm fairly sure it is called and we do need to implement it. Added a FIXME.

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/PlatformTimeRangesCocoa.mm:
```diff
@@ -0,0 +1,59 @@
+        MediaTime end = timeRanges.end(i, endValid);
+        RELEASE_ASSERT(endValid);
+
+        CMTimeRange range = PAL::CMTimeRangeMake(PAL::CMTimeMakeWithSeconds(start.toDouble(), 1000), PAL::CMTimeMakeWithSeconds(end.toDouble() - start.toDouble(), 1000));
```
I wonder if we should make the CMTimes with the MediaTime values and timescales instead of converting to a doubles and (potentially) introducing round off errors

aestes committed:
[iOS] Implement PlaybackSessionInterfaceAVKit in terms of AVMediaSource
https://bugs.webkit.org/show_bug.cgi?id=284799
rdar://problem/141601785

Reviewed by Eric Carlson.

Implemented PlaybackSessionInterfaceAVKit in terms of a content source object (WebAVContentSource)
that conforms to AVMediaSource. AVMediaSource declares a number of read-only properties that are
expected to be key-value observable, and WebAVContentSource conforms by declaring readwrite
properties that are set by various PlaybackSessionInterfaceAVKit functions. AVMediaSource methods
(e.g., -play) are implemented by calling into PlaybackSessionModel.

To avoid duplicating the method from PlaybackSessionInterfaceAVKitLegacy that converts a TimeRanges
to an NSArray of NSValues, a helper function was added to PlatformTimeRanges.

* Source/WebCore/PAL/pal/spi/cocoa/AVKitSPI.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/PlatformTimeRanges.h:
* Source/WebCore/platform/graphics/cocoa/PlatformTimeRangesCocoa.mm: Added.
(WebCore::makeNSArray):
* Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.mm:
(-[WebAVListItem initWithLocalizedTitle:]):
(-[WebAVContentSource initWithModel:]):
(-[WebAVContentSource beginScanningBackward]):
(-[WebAVContentSource beginScanningForward]):
(-[WebAVContentSource endScanningBackward]):
(-[WebAVContentSource endScanningForward]):
(-[WebAVContentSource beginScrubbing]):
(-[WebAVContentSource endScrubbing]):
(-[WebAVContentSource pause]):
(-[WebAVContentSource play]):
(-[WebAVContentSource seekTo:]):
(-[WebAVContentSource setCaptionContentInsets:]):
(-[WebAVContentSource updateCurrentAudioOption:]):
(-[WebAVContentSource updateCurrentCaptionOption:]):
(-[WebAVContentSource updateMuted:]):
(-[WebAVContentSource updateVolume:]):
(WebCore::nowPlayingMetadataObserver):
(WebCore::PlaybackSessionInterfaceAVKit::PlaybackSessionInterfaceAVKit):
(WebCore::PlaybackSessionInterfaceAVKit::durationChanged):
(WebCore::PlaybackSessionInterfaceAVKit::currentTimeChanged):
(WebCore::PlaybackSessionInterfaceAVKit::rateChanged):
(WebCore::PlaybackSessionInterfaceAVKit::seekableRangesChanged):
(WebCore::PlaybackSessionInterfaceAVKit::canPlayFastReverseChanged):
(WebCore::PlaybackSessionInterfaceAVKit::audioMediaSelectionOptionsChanged):
(WebCore::PlaybackSessionInterfaceAVKit::legibleMediaSelectionOptionsChanged):
(WebCore::PlaybackSessionInterfaceAVKit::audioMediaSelectionIndexChanged):
(WebCore::PlaybackSessionInterfaceAVKit::legibleMediaSelectionIndexChanged):
(WebCore::PlaybackSessionInterfaceAVKit::mutedChanged):
(WebCore::PlaybackSessionInterfaceAVKit::volumeChanged):
(WebCore::PlaybackSessionInterfaceAVKit::startObservingNowPlayingMetadata):
(WebCore::PlaybackSessionInterfaceAVKit::stopObservingNowPlayingMetadata):
(WebCore::PlaybackSessionInterfaceAVKit::nowPlayingMetadataChanged):
* Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKitLegacy.mm:
(WebCore::PlaybackSessionInterfaceAVKitLegacy::seekableRangesChanged):

Canonical link: https://commits.webkit.org/287992@main
